cmd/src/batch_exec.go

```diff
@@ -0,0 +1,252 @@
+		if err != nil {
+			return err
+		}
+	}
+	opts.ui.DeterminingWorkspaceCreatorTypeSuccess(workspaceCreator.Type())
+
+	// EXECUTION OF TASKS
+	coord := svc.NewCoordinator(executor.NewCoordinatorOpts{
+		Creator:       workspaceCreator,
+		CacheDir:      opts.flags.cacheDir,
+		ClearCache:    opts.flags.clearCache,
+		SkipErrors:    opts.flags.skipErrors,
+		CleanArchives: opts.flags.cleanArchives,
+		Parallelism:   opts.flags.parallelism,
+		Timeout:       opts.flags.timeout,
+		KeepLogs:      opts.flags.keepLogs,
+		TempDir:       opts.flags.tempDir,
+	})
+
+	opts.ui.CheckingCache()
+	tasks := svc.BuildTasks(ctx, batchSpec, repoWorkspaces)
+	uncachedTasks, cachedSpecs, err := coord.CheckCache(ctx, tasks)
+	if err != nil {
+		return err
+	}
+	opts.ui.CheckingCacheSuccess(len(cachedSpecs), len(uncachedTasks))
+
+	taskExecUI := opts.ui.ExecutingTasks(*verbose, opts.flags.parallelism)
+	freshSpecs, _, err := coord.Execute(ctx, uncachedTasks, batchSpec, taskExecUI)
+	if err != nil && !opts.flags.skipErrors {
+		return err
+	}
+	taskExecUI.Success()
+	if err != nil && opts.flags.skipErrors {
+		opts.ui.ExecutingTasksSkippingErrors(err)
+	}
+
+	specs := append(cachedSpecs, freshSpecs...)
+
+	ids := make([]graphql.ChangesetSpecID, len(specs))
+
+	if len(specs) > 0 {
+		opts.ui.UploadingChangesetSpecs(len(specs))
+
+		for i, spec := range specs {
+			id, err := svc.CreateChangesetSpec(ctx, spec)
+			if err != nil {
+				return err
+			}
+			ids[i] = id
+			opts.ui.UploadingChangesetSpecsProgress(i+1, len(specs))
+		}
+
+		opts.ui.UploadingChangesetSpecsSuccess(ids)
+	}
+
+	return nil
+}
+
+func loadWorkspaceExecutionInput(file string) (batcheslib.WorkspacesExecutionInput, error) {
+	var input batcheslib.WorkspacesExecutionInput
+
+	f, err := batchOpenFileFlag(&file)
+	if err != nil {
+		return input, err
+	}
+	defer f.Close()
+
+	data, err := io.ReadAll(f)
+	if err != nil {
+		return input, errors.Wrap(err, "reading workspace execution input file")
+	}
+
+	if err := json.Unmarshal(data, &input); err != nil {
+		return input, errors.Wrap(err, "unmarshaling workspace execution input file")
+	}
+
+	return input, nil
+}
+
+func convertWorkspaces(ws []*batcheslib.Workspace) []service.RepoWorkspace {
+	workspaces := make([]service.RepoWorkspace, 0, len(ws))
+
+	for _, w := range ws {
+		fileMatches := make(map[string]bool)
+		for _, path := range w.SearchResultPaths {
+			fileMatches[path] = true
+		}
+		workspaces = append(workspaces, service.RepoWorkspace{
+			Repo: &graphql.Repository{
+				ID:   w.Repository.ID,
+				Name: w.Repository.Name,
+				Branch: graphql.Branch{
+					Name: w.Branch.Name,
+					Target: graphql.Target{
+						OID: w.Branch.Target.OID,
+					},
+				},
+				ExternalRepository: struct{ ServiceType string }{ServiceType: "github"},
+				DefaultBranch: &graphql.Branch{
+					Name: w.Branch.Name,
+					Target: graphql.Target{
+						OID: w.Branch.Target.OID,
+					},
+				},
```

Comment:
Not even sure if this is required. Ideally, this field is never used directly but accessed through `Rev()` and `BaseRef()`

Reply:
Removed that and the `ExternalRepository`.